### ISSUE TITLE
Float equality checking does not issue warning

### ISSUE THREAD

**Issue by Recognition2** [rust-lang/rust#62206]
When checking  two floats for equality, the Rust compiler does not issue a warning. This is unexpected behaviour, because it is rarely necessary (nor intended behaviour) to bitwise equality-check 2 floating point numbers.

For instance, when running the following code
```rust
fn main() {
    let a: f64 = (std::f64::consts::PI * 2.).sin();
    let b: f64 = (std::f64::consts::PI * 4.).sin();

    println!("Is a equal to b? >{}", a == b);
}
```
This will print: `Is a equal to b? >false`.
Mathematically, we expect this to be true, since zero is equal to zero. 
However logically the zeroes are slightly different because of rounding errors.

This is why equality checking on floating point numbers is generally frowned upon. Instead, one should use `(a - b).abs() < epsilon` with epsilon in the region of 1e-15 (or on 32-bit representations 1e-7). This way, these miniscule differences do not cause logical errors in programs.

However, uninformed people who do not know how floats work exactly may be fooled by this. 
This is why I think the compiler should issue a warning when comparing floats, or even an error. It is very rarely necessary to bitwise compare floats. 

It is noteworthy that `gcc` does warn on this, see `-Wfloat-equal` at [the gnu gcc docs](https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html)

**Comment by jonas-schievink** [comment]
Clippy has such a lint: https://rust-lang.github.io/rust-clippy/master/index.html#float_cmp

Closing in favor of that.
